boot/systems_test.go

```diff
@@ -0,0 +1,535 @@
+}
+
+func (s *systemsSuite) SetUpTest(c *C) {
+	s.baseBootenvSuite.SetUpTest(c)
+	c.Assert(os.MkdirAll(boot.InitramfsUbuntuBootDir, 0755), IsNil)
+	c.Assert(os.MkdirAll(boot.InitramfsUbuntuSeedDir, 0755), IsNil)
+
+	restore := boot.MockSecbootResealKeys(func(params *secboot.ResealKeysParams) error { return nil })
+	s.AddCleanup(restore)
+
+	s.uc20dev = boottest.MockUC20Device("", nil)
+
+	// run kernel
+	s.runKernelBf = bootloader.NewBootFile("/var/lib/snapd/snap/pc-kernel_500.snap",
+		"kernel.efi", bootloader.RoleRunMode)
+	// seed (recovery) kernel
+	s.recoveryKernelBf = bootloader.NewBootFile("/var/lib/snapd/seed/snaps/pc-kernel_1.snap",
+		"kernel.efi", bootloader.RoleRecovery)
+}
+
+func (s *systemsSuite) TestSetTryRecoverySystemEncrypted(c *C) {
+	mockAssetsCache(c, s.rootdir, "trusted", []string{
+		"asset-asset-hash-1",
+	})
+
+	mtbl := s.mockTrustedBootloaderWithAssetAndChains(c, s.runKernelBf, s.recoveryKernelBf)
+	bootloader.Force(mtbl)
+	defer bootloader.Force(nil)
+
+	// system is encrypted
+	s.stampSealedKeys(c, s.rootdir)
+
+	modeenv := &boot.Modeenv{
+		Mode: "run",
+		// keep this comment to make old gofmt happy
+		CurrentRecoverySystems: []string{"20200825"},
+		CurrentKernels:         []string{},
+		CurrentTrustedRecoveryBootAssets: boot.BootAssetsMap{
+			"asset": []string{"asset-hash-1"},
+		},
+
+		CurrentTrustedBootAssets: boot.BootAssetsMap{
+			"asset": []string{"asset-hash-1"},
+		},
+	}
+	c.Assert(modeenv.WriteTo(""), IsNil)
+
+	var readSeedSeenLabels []string
+	restore := boot.MockSeedReadSystemEssential(func(seedDir, label string, essentialTypes []snap.Type, tm timings.Measurer) (*asserts.Model, []*seed.Snap, error) {
+		// the mock bootloader can only mock a single recovery boot
+		// chain, so pretend both seeds use the same kernel, but keep track of the labels
+		readSeedSeenLabels = append(readSeedSeenLabels, label)
+		kernelSnap := &seed.Snap{
+			Path: "/var/lib/snapd/seed/snaps/pc-kernel_1.snap",
+			SideInfo: &snap.SideInfo{
+				RealName: "pc-kernel",
+				Revision: snap.Revision{N: 1},
+			},
+		}
+		return s.uc20dev.Model(), []*seed.Snap{kernelSnap}, nil
+	})
+	defer restore()
+
+	resealCalls := 0
+	restore = boot.MockSecbootResealKeys(func(params *secboot.ResealKeysParams) error {
+		resealCalls++
+		c.Logf("params.ModelParams: %+q", params.ModelParams[0])
+		c.Logf("key files: %q", params.KeyFiles)
+
+		// bootloader variables have not been modified yet
+		c.Check(mtbl.BootVars, HasLen, 0)
+		return nil
+	})
+	defer restore()
+
+	err := boot.SetTryRecoverySystem(s.uc20dev, "1234")
+	c.Assert(err, IsNil)
+
+	c.Check(mtbl.BootVars, DeepEquals, map[string]string{
+		"try_recovery_system":    "1234",
+		"recovery_system_status": "try",
+	})
+	// run and recovery keys
+	c.Check(resealCalls, Equals, 2)
+	c.Check(readSeedSeenLabels, DeepEquals, []string{"20200825", "1234"})
+
+	modeenvRead, err := boot.ReadModeenv("")
+	c.Assert(err, IsNil)
+	c.Check(modeenvRead.CurrentRecoverySystems, DeepEquals, []string{
```

Comment:
nitpick, but I also have slight preference for doing a full DeepEquals comparison to ensure that nothing else was modified accidentally on the object

Reply:
DeepEquals pick up the `read` flag that gets set when it's read from a file.

Reply:
ah right, since we are in the boot pkg though you could use `modeenvRead.DeepEqual(exp)` I think

Reply:
Thanks for the hint, I've updated the tests.